opencl/btx_clinterval_callbacks.cpp

```diff
@@ -0,0 +1,354 @@
+                                              const char *event_class_name,
+                                              const char *hostname, int64_t vpid,
+                                              uint64_t vtid, cl_device_id in_device) {
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t key{hostname, vpid, vtid};
+  state->entry_state.set_data(key, reinterpret_cast<thapi_device_id>(in_device));
+}
+
+
+static void create_sub_devices_exit_callback(void *btx_handle, void *usr_data, int64_t ts,
+                                             const char *event_class_name,
+                                             const char *hostname, int64_t vpid,
+                                             uint64_t vtid, cl_int errcode_ret_val,
+                                             cl_uint num_devices_ret_val,
+                                             cl_device_id *out_devices_vals) {
+  auto state = static_cast<data_t *>(usr_data);
+  if ((out_devices_vals != nullptr) && (errcode_ret_val == CL_SUCCESS)) {
+    hpt_t key{hostname, vpid, vtid};
+    auto in_device = state->entry_state.get_data<thapi_device_id>(key);
+    const thapi_device_id root_device = state->device_to_root_device[
+                                                  hp_device_t(hostname, vpid, in_device)];
+    for (unsigned int i=0; i < num_devices_ret_val; i++ ) {
+      const thapi_device_id d = reinterpret_cast<thapi_device_id>(out_devices_vals[i]);
+      state->device_to_root_device[hp_device_t(hostname, vpid, d)] = root_device;
+    }
+  }
+}
+
+static void create_command_queue_entry_callback(void *btx_handle, void *usr_data,
+                                                int64_t ts, const char *event_class_name,
+                                                const char *hostname, int64_t vpid,
+                                                uint64_t vtid, cl_device_id device) {
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t key{hostname, vpid, vtid};
+  state->entry_state.set_data(key, reinterpret_cast<thapi_device_id>(device));
+}
+
+static void create_command_queue_exit_callback(void *btx_handle, void *usr_data,
+                                               int64_t ts, const char *event_class_name,
+                                               const char *hostname, int64_t vpid,
+                                               uint64_t vtid,
+                                               cl_command_queue command_queue,
+                                               cl_int errcode_ret_val) {
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t key{hostname, vpid, vtid};
+  auto entry_device = state->entry_state.get_data<thapi_device_id>(key);
+  const auto root_device = state->device_to_root_device[
+                                              hp_device_t(hostname, vpid, entry_device)];
+  hp_command_queue_t cq_key{hostname, vpid, command_queue};
+  if (root_device != 0) {
```

Comment:
I think it cannot be `0` As we always add it to the map

Reply:
The map is set on `clGetDeviceIDs` and in `clCreateSubDevices`. I think the idea is that if you are using a top level device, you may never call these so the map would not be set and it would be 0, and would be a root device. This was copied from the old erb code.

Reply:
Looking more at OpenCL API, cl_device_id is a pointer to a struct, so you can't easily skip the step of `clGetDeviceIDs` or `clCreateSubDevices`. I am unsure however if there is another API or something we are not covering, so hesitant to remove the if branch since it *might* break something and was in the old code.

Reply:
Agree, lets keep it for now! It's just a porting after all :) We can do refactoring in another pr